PassthroughConnection.close after PassthroughServer.stop throws AssertionError instead of IOException

If `PassthroughConnection.close()` is called after `PassthroughServer.stop()` is called, an `AssertionError` is thrown instead of an `IOException` -- should be an `IOException`.

----
Sample stack trace:

````
java.lang.AssertionError: Case must be true
    at org.terracotta.passthrough.Assert.assertTrue(Assert.java:41)
    at org.terracotta.passthrough.PassthroughServerProcess.sendMessageToServer(PassthroughServerProcess.java:316)
    at org.terracotta.passthrough.PassthroughConnectionState.createAndSend(PassthroughConnectionState.java:79)
    at org.terracotta.passthrough.PassthroughConnectionState.sendNormal(PassthroughConnectionState.java:65)
    at org.terracotta.passthrough.PassthroughConnection.invokeAndWait(PassthroughConnection.java:138)
    at org.terracotta.passthrough.PassthroughConnection.sendUnexpectedCloseMessage(PassthroughConnection.java:367)
    at org.terracotta.passthrough.PassthroughConnection.close(PassthroughConnection.java:324)
    at org.ehcache.clustered.client.internal.UnitTestConnectionService.remove(UnitTestConnectionService.java:169)
    at org.ehcache.clustered.client.internal.UnitTestConnectionService.remove(UnitTestConnectionService.java:189)
    at org.ehcache.clustered.client.internal.service.DefaultClusteringServiceTest.removePassthroughServer(DefaultClusteringServiceTest.java:99)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:112)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:56)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
    at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
    at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
    at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
    at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:364)
    at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
    at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:662)
````

--- a/entity-test-lib/src/main/java/org/terracotta/passthrough/PassthroughConnection.java
+++ b/entity-test-lib/src/main/java/org/terracotta/passthrough/PassthroughConnection.java
@@ -129,11 +129,9 @@ public class PassthroughConnection implements Connection {
   }
 
   private PassthroughWait invokeAndWait(PassthroughMessage message, boolean shouldWaitForSent, boolean shouldWaitForReceived, boolean shouldWaitForCompleted, boolean shouldWaitForRetired, boolean forceGetToBlockOnRetire) {
-    // If we have already disconnected, fail.
+    // If we have already disconnected, fail with IllegalStateException (this is consistent with the double-close case).
     if (!this.isRunning) {
-      // TODO:  Determine a more appropriate exception if tests come to test this path.
-      // (for this, this is just to make testing easier)
-      throw new RuntimeException("Connection not open");
+      throw new IllegalStateException("Connection already closed");
     }
     PassthroughWait waiter = this.connectionState.sendNormal(this, message, shouldWaitForSent, shouldWaitForReceived, shouldWaitForCompleted, shouldWaitForRetired, forceGetToBlockOnRetire);
     if (Thread.currentThread() == clientThread) {
@@ -313,22 +311,27 @@ public class PassthroughConnection implements Connection {
 
   @Override
   public void close() {
-    // Note that the caller may not know if we were already closed by another test trying to simulate unexpected disconnect situations so check this.
+    // We expect a double-close to throw IllegalStateException so check if we are running.
     if (this.isRunning) {
-      // Tell each our still-connected end-points to disconnect from the server.
-      // (we use a message for this, since it keeps the flow obvious, but we should probably use something more jarring)
-      for (PassthroughEntityClientEndpoint<?, ?> endpoint : this.localEndpoints.values()) {
-        // Ask the end-point to create the message (only has it the information regarding what entity this is, etc).
-        PassthroughMessage releaseMessage = endpoint.createUnexpectedReleaseMessage();
-        // Send the message.
-        sendUnexpectedCloseMessage(releaseMessage);
-      }
-      
-      // We also need to drop all locks.
-      for (PassthroughEntityTuple lockedEntity : this.writeLockedEntities) {
-        PassthroughMessage message = PassthroughMessageCodec.createDropWriteLockMessage(lockedEntity.entityClassName, lockedEntity.entityName);
-        // Send the message.
-        sendUnexpectedCloseMessage(message);
+      // It is possible that the server already shut down and will throw IllegalStateException so catch that, here.
+      try {
+        // Tell each our still-connected end-points to disconnect from the server.
+        // (we use a message for this, since it keeps the flow obvious, but we should probably use something more jarring)
+        for (PassthroughEntityClientEndpoint<?, ?> endpoint : this.localEndpoints.values()) {
+          // Ask the end-point to create the message (only has it the information regarding what entity this is, etc).
+          PassthroughMessage releaseMessage = endpoint.createUnexpectedReleaseMessage();
+          // Send the message.
+          sendUnexpectedCloseMessage(releaseMessage);
+        }
+        
+        // We also need to drop all locks.
+        for (PassthroughEntityTuple lockedEntity : this.writeLockedEntities) {
+          PassthroughMessage message = PassthroughMessageCodec.createDropWriteLockMessage(lockedEntity.entityClassName, lockedEntity.entityName);
+          // Send the message.
+          sendUnexpectedCloseMessage(message);
+        }
+      } catch (IllegalStateException e) {
+        // Ignore this - it just means the server is shut down so we don't need to send them any messages.
       }
      
       // We are going to stop processing messages so set us not running and stop our thread.
@@ -353,6 +356,8 @@ public class PassthroughConnection implements Connection {
       // We might as well drop the references from our tracking, also, since they can't reasonably be used.
       this.localEndpoints.clear();
       this.writeLockedEntities.clear();
+    } else {
+      throw new IllegalStateException("Connection already closed");
     }
   }
 

--- a/entity-test-lib/src/main/java/org/terracotta/passthrough/PassthroughServerProcess.java
+++ b/entity-test-lib/src/main/java/org/terracotta/passthrough/PassthroughServerProcess.java
@@ -313,7 +313,11 @@ public class PassthroughServerProcess implements MessageHandler, PassthroughDump
   }
 
   public synchronized void sendMessageToServer(final PassthroughConnection sender, byte[] message) {
-    Assert.assertTrue(this.isRunning);
+    // If the server shut down, throw IllegalStateException
+    if (!this.isRunning) {
+      throw new IllegalStateException("Connection already closed");
+    }
+
     MessageContainer container = new MessageContainer();
     container.sender = new IMessageSenderWrapper() {
       @Override

